Enhancements/wish lists pr

MatteoPiovanelli-Laser opened the pull request:
#117 PR on wish lists.

This is partially a WIP. There are still some UI things we are working on, especially regarding the "add to wishlist" menu from products' display views. We are also looking into notusing javascript for the view of the "add to wishlist" shapes.

The following functionality is in place for authenticated users. Unauthenticated users get redirected to the logon page.

- From a product's page:
  - Add the product, along wiht selected attributes, to an existing wishlist.
  - Add the product, along with selected attributes, to a new wishlist.
  - Adding a product to a wishlist redirects to the page for that wishlist.
- From the `/wishlists` route:
  - See the details of any owned wishlist.
    - For each item in the wishlist, add it to the shopping cart or remove it from the wishlist.
  - Create a new wishlist.
  - Update wishlists, by changing their titles, selecting which wishlist is the default one, deleting wishlists.

bleroy commented:
Should I review now, or should I wait for the work to be complete?

MatteoPiovanelli-Laser commented:
If you have time, I'd appreciate it if you could give it a quick look. We are making some changes for the sake of clarity, but maybe you will spot something we oversaw.

MatteoPiovanelli-Laser commented:
@HermesSbicego-Laser and I updated things a bit.
We simplified the UI, moving away from using HTML5 css3 modals everywhere, because that could have easily become a nightmare when trying to alternate from a theme.
We also fixed a few bugs here and there that we spotted.

bleroy commented on ViewModels/WishListsIndexViewModel.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Nwazet.Commerce.ViewModels {
+    public class WishListsIndexViewModel {
+        //public dynamic CreateShape { get; set; }
```
Please remove commented out code.

MatteoPiovanelli-Laser replied:
ok

bleroy commented on ViewModels/WishListsIndexViewModel.cs:
```diff
@@ -0,0 +1,17 @@
+        //public dynamic CreateShape { get; set; }
+        //public dynamic SettingsShape { get; set; }
+        public IEnumerable<WishListListPart> WishLists { get; set; }
+        public WishListListPart WishList { get; set; }
```
Unclear what this is for. The name should probably be more specific.

MatteoPiovanelli-Laser replied:
I got rid of the class entirely, because I was not using it anymore

bleroy commented on Views/Parts/Product.AddToCartFromWishlist.cshtml:
```diff
@@ -0,0 +1,35 @@
+    Style.Include("wishlists.nwazet-commerce.css", "wishlists.nwazet-commerce.min.css");
+
+    var productId = (int)Model.ProductId;
+    int quantityValue = Model.MinimumOrderQuantity > 1 ? Model.MinimumOrderQuantity : 1;
```
`Math.Min(1, Model.MinimumOrderQuantity)`?

MatteoPiovanelli-Laser replied:
`Math.Max` but yeah

bleroy commented on Views/Parts/Product.AddToWishlistButton.cshtml:
```diff
@@ -0,0 +1,60 @@
+@using (Script.Foot()) {
+    <script type="text/javascript">
+        jQuery(function ($) {
+            $('#wishlist-add').click(function () {
```
There's a lot of elements in here that use a statically assigned, fixed id. Even if we think there will ever be one of those on the page, this could be made cleaner using `Id` helpers that MVC provides. Also, whenever you have a script like that, it should be in a separate script file. Inline scripts in the view should only be initializing properties to dynamic values.

bleroy commented:
Wow, that's a pretty big feature. Thanks for the contribution! I had lots of comments (because that's a lot of code).

I'm also still about the use of the word "element": we have something else in Orchard that is also called, confusingly enough, an element, and it seems like the shopping carts uses "item" for something similar.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+                return new HttpUnauthorizedResult();
+            }
+            var wishLists = _wishListServices.GetWishLists(user);
+            var selectedList = wishLists.SingleOrDefault(wp => wp.ContentItem.Id == id);
```
yep. This was code left over from the prvious version of this controller

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+
+            _shoppingCart.Add(productId, quantity, productattributes);
+
+            _workflowManager.TriggerEvent("CartUpdated",
```
I am working on a separate branch on cart events.
Stuff like adding/removing items will have their own events.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+        public ActionResult Index(int id = 0) {
+            var user = _wca.GetContext().CurrentUser;
+            if (user == null) {
+                return new HttpUnauthorizedResult();
```
`[Authorize]` does exactly that very same thing.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            var wishLists = _wishListServices.GetWishLists(user);
+            var selectedList = wishLists.SingleOrDefault(wp => wp.ContentItem.Id == id);
+            if (selectedList == null) {
+                selectedList = wishLists.SingleOrDefault(wp => wp.IsDefault);
```
ok

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+                selectedList = wishLists.SingleOrDefault(wp => wp.IsDefault);
+            }
+            return View(_contentManager.BuildDisplay(selectedList));
+        }
```
ok

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+        }
+        [Authorize]
+        [HttpPost, ActionName("Create")]
+        public ActionResult CreatePost(string new_wishlist_title, int productid = 0) {
```
ok

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+
+
+        [HttpPost]
+        public ActionResult CreateWishList(string new_wishlist_title, int productid = 0) {
```
By closer, you mean actually moving this method up in the code?

bleroy replied:
yes

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            wishlistId = wishList.ContentItem.Id;
+            //add product to wishlist
+            if (productid > 0) {
+                var productPart = _contentManager.Get<ProductPart>(productid);
```
added null-check

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            var wishList = _wishListServices.GetWishList(user, wishListid);
+
+            if (productid > 0) {
+                var productPart = _contentManager.Get<ProductPart>(productid);
```
added null-check

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+
+        [HttpPost]
+        public ActionResult AddToWishList(int wishListid, int productid) {
+            var user = _wca.GetContext().CurrentUser;
```
Honestly, `[Authorize]` really does everything I was trying to do here.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+        }
+
+        [HttpPost]
+        public ActionResult RemoveFromWishList(int wishListid, int elementId) {
```
WishListElements are now WishListItems

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            //get selected wishlist
+            var wishList = _wishListServices.GetWishList(user, wishListid);
+
+            if (wishList.ContentItem.Id == wishListid) { //GetWIshList may return the default wish list
```
I added a `TryGetWishList` method that returns true if the wishlist is found, false if it had to fallback on the default.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            //1. Read what should change from the form
+            var form = HttpContext.Request.Form;
+            //1.1 Get the id of the new default wish list
+            var defaultId = int.Parse(form["is-default-wishlist"]);
```
Because I am dumb

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            if (user == null) {
+                return new HttpUnauthorizedResult();
+            }
+            var wishLists = _wishListServices.GetWishLists(user);
```
moved

MatteoPiovanelli-Laser commented on Drivers/ProductPartWishListDriver.cs:
```diff
@@ -0,0 +1,60 @@
+            var user = _wca.GetContext().CurrentUser;
+            //usually, wishlists only apply to authenticated users.
+            //however, we are still going to show the "Add to list" functionalities, to drive sign-up / login
+            var wishLists = new List<WishListListPart>();
```
ok

MatteoPiovanelli-Laser commented on Drivers/WishListListPartDriver.cs:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        protected override DriverResult Display(WishListListPart part, string displayType, dynamic shapeHelper) {
+            var shapes = new List<DriverResult>();
```
ok

MatteoPiovanelli-Laser commented on Models/WishListListPart.cs:
```diff
@@ -0,0 +1,44 @@
+            get {
+                return _serializedItemIds
+                    .Split(separator, StringSplitOptions.RemoveEmptyEntries)
+                    .Select(frag => int.Parse(frag))
```
fixed

MatteoPiovanelli-Laser replied:
I took this from `ContentPickerField`, only I do `Retrieve` on my record rather than going through field storage.

bleroy replied:
ok, well, `ContentPickerField` probably shouldn't be doing that, but it's likely too late to change now without breaking people. Here, we have an opportunity to do it better.

MatteoPiovanelli-Laser commented on Models/WishListListPart.cs:
```diff
@@ -0,0 +1,44 @@
+                    .ToArray();
+            }
+            set {
+                _serializedItemIds = "{" + string.Join("},{", value) + "}";
```
I took this form `ContentPickerField` and I assumed there was a reason for it I could not see. A comment from `ContentPickerField.cs`:
`// use {1},{2} format so it can be filtered with delimiters`

bleroy replied:
I don't understand what that means. I'll ask Sébastien.

bleroy replied:
OK, I asked, and he confirmed that the serialization to a string is bad, and he's not doing it any more on more recent code. The braces are there because he's doing `contains` searches on the contents of that field, which in itself is pretty messed up. That should use an index, rather than a search on the contents of the infoset. But here, we're never going to search on that field anyway, so there's no reason to do the weird serialization and braces.

MatteoPiovanelli-Laser replied:
I got rid of the string entirely. Since I am storing the Id of the WishList in the record for the item, I figure I don't really need the string anyway.

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+               .Join<TitlePartRecord>()
+               .OrderBy(p => p.Title);
+            if (max > 0) {
+                wishLists.AddRange(wishListsQuery.Slice(max - 1)); //-1, because the default wishlist is there already
```
I agree. I just never use `Concat`. Doesn't ever occur to me it's even there.

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+                .Where(wllp => wllp.IsDefault)
+                .Join<CommonPartRecord>()
+                .Where(cpr => cpr.OwnerId == user.Id) //match owner
+                .Slice(1).FirstOrDefault();
```
I am not aware of `IContentQuery` having its own `FirstOrDefault` method, hence I am using `slice` because I only really want one result, and then I take the first item in the `IEnumerable` that returns.
Is that extension actually somewhere?

bleroy replied:
Oh, I see, you're right. This is weird but I don't see a better way.

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+            //check whether the product is in the wishlist already
+            if (!ItemIsInWishlist(wishlist, item)) {
+                //create a new wishlist element and add it
+                var newELement = _contentManager.New<WishListElementPart>("WishListItem");
```
typo.

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+                newELement.Item = item;
+                _contentManager.Create(newELement.ContentItem);
+                //add to list
+                var elementIds = wishlist.Ids?.ToList();
```
That was a leftover from a bug fix that I did earlier on `WishListListPart`

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+                throw new ArgumentNullException("user");
+            }
+            if (wishlist.ContentItem.As<CommonPart>().Owner != user) {
+                throw new ArgumentException(T("Users may only edit wishlists they own.").Text);
```
ok

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+                ShoppingCartItem.ItemsAreEqual(wel.Item, item));
+        }
+
+        public void RemoveElementFromWishlist(IUser user, WishListListPart wishlist, int elementId) {
```
ok

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+            var elementId = elementPart.ContentItem.Id;
+            if (wishlist.Ids.Contains(elementId)) {
+                var elementIds = wishlist.Ids.ToList().Where(id => id != elementId);
+                wishlist.Ids = elementIds.ToArray();
```
The `ToList` is completely useless there.
The `ToArray` is because the property I am assigning to is an `int []`

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+            _contentManager.Destroy(elementPart.ContentItem); //hard delete
+        }
+
+        public void DeleteWishlist(IUser user, WishListListPart wishlist) {
```
ok

MatteoPiovanelli-Laser commented on Services/WishListServices.cs:
```diff
@@ -0,0 +1,312 @@
+            _contentManager.Destroy(wishlist.ContentItem);
+        }
+
+        public dynamic CreateShape(IUser user, ProductPart product = null) {
```
I would agree, however to create these shapes I am using injected services, and afaik I can't inject them in a static class. I guess I could resolve those in a static helper class, but I'd rather not.
The idea with these methods is that different extensions can provide their own contribution to the UI shapes. E.g.: an extension allowing wishlists to be made public would likely have to add to the the settings view, to allow changing a wishlist between public/private.

bleroy replied:
Oh well, I said static, but it doesn't have to be, if it would be too many parameters to pass to the methods. I would still separate into something like a `WishListUIService`.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+        }
+
+        [HttpPost]
+        public ActionResult AddToCart(int productId, int quantity = 1) {
```
Yes, but not quite.
`ShoppingCartController.Add` expects a form in its request to extract the attribute values, that form being "created" by `ProductPartDriver` in `Parts/Product.AddButton.cshtml` and then filled by the user.
In this case, I would not have to fill the form, because all the attribute values have already been selected before the item is added to a wishlist. for this reason I am using a different form.

I guess I could move this action to the `ShoppingCartController`, but I would not try to "fold" what it does in the same `Add` method we use from the product's page.

MatteoPiovanelli-Laser replied:
BTW, I changed this to use model binding rather than the form.

MatteoPiovanelli-Laser replied:
even then, however, if I hit F12, change the values from that form, and then click "Add to cart", the product that is added has the edited values.
e.g.:
 - my wishlist has a ProductX in Size 56 (attribute and value)
 - I hit F12 and change the valu of the hidden to 52
 - I click on the "Add to cart" button
 - The product added to the cart has Size 52

MatteoPiovanelli-Laser replied:
I am also working on adding a permission layer, so I don't always have to pass the user around (unless I need a user)

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+        public ActionResult AddToCart(int productId, int quantity = 1) {
+
+            //read attributes from the form
+            var form = HttpContext.Request.Form;
```
not a form any more

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            var form = HttpContext.Request.Form;
+            var productattributes = form.AllKeys
+                .Where(key => key.StartsWith("attributeKey"))
+                .Select(key => int.Parse(form[key]))
```
using model binding now, and there is a bit of validation.

MatteoPiovanelli-Laser commented on Controllers/WishListsController.cs:
```diff
@@ -0,0 +1,256 @@
+            return RedirectToAction("Index", new { id = wishListid });
+        }
+
+        private Dictionary<int, ProductAttributeValueExtended> ParseProductAttributes() {
```
I copied this verbatim from the `ShoppingCartController`

bleroy replied:
mmm, that's concerning. Would you mind opening an issue so that I can review that there?

MatteoPiovanelli-Laser replied:
ok

MatteoPiovanelli-Laser commented:
Thanks for taking the time to review the code.

I made most changes you recommended (update to the PR will be out in a few minutes). Where I replied "ok" to your comments it was so that I would see that I had worked on that point.

A few things are still not finalized:

- `WishListController.AddToCart`: it uses model binding now, and verifies things about the wish list before adding the product to the cart. 
- I will add a permission provider to handle the ability to see/update wish lists. It will be based on a service (whose default implementation will be such that people can only interact with their own wishlists), allowing to implement different things (e.g. public wishlists, gift lists...). `IWishlistServices` will change to reflect not always needing the user as a parameter.
- Cart events are on a separate branch I am working on.
- Attributes being read from the form: I copied that code from the `ShoppingCartController`. It may be worth to opne a seperate issue to see if it's possible to convert that so it does no use the form, but I feel that since we cannot know what will be in the extensions, we may be out of luck here. A rewrite may be possible such that extensions have a common portion that allows identificaition, and then they process the form themselves if needed, so that we may remove it form the controller. It may be possible to bind a dictionary similarly to what I did for `WIshListController.AddToCart`.
- serialization/deserialization of `WishListListPart.Ids`: it's basically the same as `ContentPickerField.Ids`, but in my own record rather than `StringFieldIndexRecord`.
- Service methods to provide shapes: I am using injected services and providers, to make the shapes extensible.

MatteoPiovanelli-Laser commented:
Il the latest PR there are more fixes and features:

- I moved the methods to create those shapes to a separate `IWishListsUIServices`.
- I got rid of that string in `WishListListPart` (and `WishListListPartRecord`): since in `WishListItemPartRecord` I have the id of the "container" wish list anyway, the list of items in a wish list is taken by querying on that. I think this also makes things cleaner, because this way when the items in a wish list are updated (add/remove) there is only one place that stores the info, rather than two that need to be kept in synch.
- I added permissions related to wishlists. I added `WishListsAuthorizationEventHandler`. The default behaviour is "wish lists are all private at all times". See `IWishListsPermissionsService` and its provided implementation.

bleroy commented on Controllers/WishListsController.cs:
```diff
@@ -19,44 +20,64 @@ namespace Nwazet.Commerce.Controllers {
         [OutputCache(Duration = 0)]
+        public ActionResult Index(string uName, int id = 0) {
+            //the uName string is there to allow accessing other user's wishlists
+            var user = string.IsNullOrWhiteSpace(uName) ? _wca.GetContext().CurrentUser
```
I don't get it. What's the point? The `user` variable is not used as far as I can tell. What determines who can view whose wishlists?

bleroy replied:
By the way, I think there should be ways to share a wishlist.

MatteoPiovanelli-Laser replied:
1. `user`: my bad, I forgot to pass it to `TryGetWishList`.
2. The way the implementation is right now, wish lists are private. The actual authorizer goes through `WishListsAuthorizationEventHandler`, that in turn calls to methods from `IWishListsPermissionsService`. The default implementation I am providing implements the concept of private wish lists. Public or shareable wish lists can be provided by a separate feature extending on this. e.g., to implement always public wish lists it's enough to implement `IWishListsPermissionsService` with empty methods.

bleroy commented on Controllers/WishListsController.cs:
```diff
@@ -19,44 +20,64 @@ namespace Nwazet.Commerce.Controllers {
+            WishListListPart selectedList;
+            if (_wishListServices.TryGetWishList(out selectedList, id)) {
+                if (!_orchardServices.Authorizer.Authorize(WishListPermissions.ViewWishLists, selectedList))
+                    return RedirectToAction("Index", new { id = 0 }); //redirect to own wish lists
```
If you don't have permissions to see any list, this looks like it will be an infinite redirect loop.

MatteoPiovanelli-Laser replied:
Yes.
My assumption is that a user will always be allowed to see their wish lists, and the way the services are implemented, every user will always have at least 1 wish list.
I don't see a reasonable use case where a user would not be allowed to see their own wish lists.
Of course it's possible to implement the services so that users may not have any wishlist. I don't see a case where that would make sense.
Anyway, I can handle that: I am thinking of adding a check and returning an error page when the user has no wishlists they can see. Does that make sense?

bleroy replied:
It wouldn't necessarily a sane state, but it's one that is possible, so we should fail gracefully. Infinite redirects are never good ;)

bleroy commented on Controllers/WishListsController.cs:
```diff
@@ -19,44 +20,64 @@ namespace Nwazet.Commerce.Controllers {
         private const string ExtensionPrefix = "ext.";
 
         [OutputCache(Duration = 0)]
+        public ActionResult Index(string uName, int id = 0) {
```
Shouldn't "Index" be a view of all your wish lists, and we'd have a separate action to view a specific list?

MatteoPiovanelli-Laser replied:
right now it's both, controlled by placement.info. The `Display` method from `WishListListPartDriver` returns 3 shapes: the details of the list, the list of the lists, the "actions" on lists.

MatteoPiovanelli-Laser replied:
done

MatteoPiovanelli-Laser commented:
I updated the PR.

- The views returned by the `Index` method. This way it's easy to use placement info to only show what we want. E.g. something like this would result in in showing the detail only when "asking" for a specific wish list, and only the list of wish lists on the "generic" path to the action:
```
<Match Path="/wishlists">
    <Place Parts_ListOfWishLists="Content:1"/>
    <Place Parts_WishListList="-"/>
    <Place Parts_WishListsActions="-"/>
</Match>
<Match Path="/wishlists/*">
    <Place Parts_ListOfWishLists="-"/>
    <Place Parts_WishListList="Content:1"/>
    <Place Parts_WishListsActions="-"/>
</Match>
```

- I am handling the case where a user has no wish list they are allowed to see with a 404.
- public wish lists: since this is a pretty big feature already, I'd rather leave those considerations for separate/additional features later on, especially since those can be become as complex as we wish.

bleroy commented:
Thanks. There are still conflicts to solve however.

MatteoPiovanelli-Laser commented:
I fixed the merge conflicts, and also used the new cart events in the `WishListsController.AddToCart` method

bleroy commented:
Thanks!

MatteoPiovanelli-Laser commented:
thanks for reviewing this and merging it in,
